crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -348,15 +417,23 @@ impl<'db> InnerIntersectionBuilder<'db> {
                 // simplify the representation.
                 self.add_positive(db, ty);
             }
-            // bool & ~Literal[True] = Literal[False]
-            // bool & ~AlwaysTruthy = Literal[False]
-            Type::BooleanLiteral(_) | Type::AlwaysFalsy | Type::AlwaysTruthy
-                if self.positive.contains(&KnownClass::Bool.to_instance(db)) =>
-            {
-                *self = Self::default();
-                self.positive.insert(Type::BooleanLiteral(
-                    new_negative.bool(db) != Truthiness::AlwaysTrue,
-                ));
+            // `bool & ~AlwaysTruthy` -> `bool & Literal[False]`
+            // `bool & ~Literal[True]` -> `bool & Literal[False]`
+            Type::AlwaysTruthy | Type::BooleanLiteral(true) if contains_bool() => {
+                self.add_positive(db, Type::BooleanLiteral(false));
+            }
```

Comment:
I suspect the codspeed improvement is because we're now doing fewer Salsa lookups eagerly in places like this